code/modules/mob/living/simple_animal/hostile/pulse_demon/pulsedemon.dm

```diff
@@ -118,28 +119,57 @@
 	if(statpanel("Status"))
 		stat(null, text("Charge stored: [charge]W"))
 		stat(null, text("Max charge stored: [maxcharge]W"))
-
-/mob/living/simple_animal/hostile/pulse_demon/Life()
-	// Add the regen rate unless it puts us over max health, then just cap it off
+		
+/mob/living/simple_animal/hostile/pulse_demon/proc/power_lost()
+	health -= health_drain_rate
+	if(!powerloss_alerted)
+		to_chat(src, "You have lost power!")
+		powerloss_alerted = 1
+		//TODO add a sound
+		
+/mob/living/simple_animal/hostile/pulse_demon/proc/update_glow()
+	var/range = 1.5 
+	if(charge <= 50000) //why isn't there a log function, i guess i could have used len(str()) but that's fucking stupid
+		range = 1.5
+	else if(charge <= 100000)
+		range = 2
+	else if(charge <= 200000)
+		range = 2.5
+	else if(charge <= 400000)
+		range = 3
+	else
+		range = 3.5
+	set_light(range, 2, "#bbbb00")
+	
+/mob/living/simple_animal/hostile/pulse_demon/proc/power_restored()
 	var/health_to_add = maxHealth - health < health_regen_rate ? maxHealth - health : health_regen_rate
+	if(health < maxHealth)
+		health += health_to_add
+	if(powerloss_alerted)
+		to_chat(src, "Power restored.")
+		powerloss_alerted = 0
+		//TODO add a sound
+	
+/mob/living/simple_animal/hostile/pulse_demon/Life()
+	update_glow()
 	if(current_cable)
 		if(current_cable.avail() < amount_per_regen) // Drain our health if powernet is dead, otherwise drain powernet
-			health -= health_drain_rate
-		else if(health < maxHealth && draining)
+			power_lost()
+		else
+			power_restored()
 			current_cable.add_load(amount_per_regen, POWER_PRIORITY_BYPASS)
-			health += health_to_add
 	else if(current_power)
 		if(istype(current_power,/obj/machinery/power/battery) && draining)
 			var/obj/machinery/power/battery/current_battery = current_power
 			suckBattery(current_battery)
 		else if(istype(current_power,/obj/machinery/power/apc) && draining)
 			var/obj/machinery/power/apc/current_apc = current_power
 			drainAPC(current_apc)
-		if(current_power.avail() < amount_per_regen) // Drain our health if powernet is dead, otherwise drain powernet
-			health -= health_drain_rate
-		else if(health < maxHealth && draining)
-			current_cable.add_load(amount_per_regen, POWER_PRIORITY_BYPASS)
-			health += health_to_add
+		if(current_power.avail() < amount_per_regen)
+			power_lost()
+		else
+			power_restored()
+			//current_cable.add_load(amount_per_regen, POWER_PRIORITY_BYPASS) //TODO fix this, current cable is null and runtimes; i have no idea where else to draw power from
```

Comment:
we're inside a power machine at this point aren't we? so why not get its power source?

Reply:
i'm not sure how, i kept getting runtimes

Reply:
First try to get the cable. If that's null, use get_cell() to get the powercell. This will cover the APC situation. If that is also null, just skip draining because clearly there is nothing to drain. However obviously the demon won't be able to charge in this case.

Reply:
guess this might also count as balance - should a demon in an empty powernet heal while sucking an APC's power cell?
how about if you're in a charged SMES that's had its output cable snipped? 

in both these cases current_machine.avail() returns 0

Reply:
Oh, that's fair. In this case then, if you have nothing reasonable to suck from because it's not part of a powernet, it's fine to just not drain anything. It would be reasonable to implement either or both of the things you described as its own PR so it would be more atomic -- if you want to pursue either at all.

In this case, I was just misunderstanding how PD works. Still haven't ever been picked for the role.

Reply:
@kane-f